poseidon/poseidon.go

```diff
@@ -0,0 +1,184 @@
+package poseidon
+
+import (
+	"bytes"
+	"errors"
+	"math/big"
+	"strconv"
+
+	"github.com/iden3/go-iden3-crypto/field"
+	"golang.org/x/crypto/blake2b"
+)
+
+const SEED = "poseidon"
+const NROUNDSF = 8
+const NROUNDSP = 57
+const T = 6
+
+var constants = generateConstantsData()
+
+type constantsData struct {
+	fqR field.Fq
+	c   []*big.Int
+	m   [][]*big.Int
+}
+
+// checkBigIntInField checks if given big.Int fits in a Field R element
+func checkBigIntInField(a *big.Int, q *big.Int) bool {
+	if a.Cmp(q) != -1 {
+		return false
+	}
+	return true
+}
+
+// checkBigIntArrayInField checks if given big.Int fits in a Field R element
+func checkBigIntArrayInField(arr []*big.Int, q *big.Int) bool {
+	for _, a := range arr {
+		if !checkBigIntInField(a, q) {
+			return false
+		}
+	}
+	return true
+}
+
+func generateConstantsData() constantsData {
+	var constants constantsData
+
+	r, ok := new(big.Int).SetString("21888242871839275222246405745257275088548364400416034343698204186575808495617", 10)
+	if !ok {
+
+	}
+	fqR := field.NewFq(r)
+	constants.fqR = fqR
+	constants.c = getPseudoRandom(fqR, SEED+"_constants", big.NewInt(int64(NROUNDSF+NROUNDSP)))
+	constants.m = getMDS(fqR)
+
+	return constants
+}
+
+func getPseudoRandom(fqR field.Fq, seed string, n *big.Int) []*big.Int {
+	var res []*big.Int
+	hash := blake2b.Sum256([]byte(seed))
+	for big.NewInt(int64(len(res))).Cmp(n) == -1 { // res < n
+		newN := fqR.Affine(leByteArrayToBigInt(fqR, hash[:]))
+		res = append(res, newN)
+		hash = blake2b.Sum256(hash[:])
+	}
+	return res
+}
+
+func leByteArrayToBigInt(fqR field.Fq, b []byte) *big.Int {
```

Comment:
Why do you need `fqR` as an input argument?  Can't you set `res := big.NewInt(0)`?